sdk/cosmosdb/cosmos/src/encryption/EncryptionKeyResolver/AzureKeyVaultEncryptionKeyResolver.ts

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import type { TokenCredential } from "@azure/core-auth";
+import type { EncryptionKeyResolver } from "./EncryptionKeyResolver";
+import type { KeyWrapAlgorithm } from "@azure/keyvault-keys";
+import { KeyClient } from "@azure/keyvault-keys";
+
+/**
+ * Implementation of EncryptionKeyResolver that uses Azure Key Vault for customer managed keys.
+ */
+export class AzureKeyVaultEncryptionKeyResolver implements EncryptionKeyResolver {
+  private credentials: TokenCredential;
+
+  constructor(credentials: TokenCredential) {
+    this.credentials = credentials;
+  }
+  /**
+   * wraps the given key using the specified key encryption key path and algorithm.
+   * @param encryptionKeyId - path to the customer managed key to be used for wrapping.
+   * @param algorithm - algorithm to be used for wrapping.
+   * @param key - dek to be wrapped.
+   * @returns wrapped DEK.
+   */
+  public async wrapKey(encryptionKeyId: string, algorithm: string, key: Buffer): Promise<Buffer> {
+    const origin = this.getOrigin(encryptionKeyId);
+    const keyClient = new KeyClient(origin, this.credentials);
+    const [keyName, keyVersion] = this.getKeyDetails(encryptionKeyId);
+    const cryptographyClient = keyClient.getCryptographyClient(keyName, {
+      keyVersion: keyVersion,
+    });
+    const res = await cryptographyClient.wrapKey(algorithm as KeyWrapAlgorithm, key);
+    return Buffer.from(res.result);
+  }
+  /**
+   * Unwraps the given wrapped key using the specified key encryption key path and algorithm.
+   * @param encryptionKeyId - path to the customer managed key to be used for unwrapping.
+   * @param algorithm - algorithm to be used for unwrapping.
+   * @param wrappedKey - wrapped DEK.
+   * @returns unwrapped DEK.
+   */
+  public async unwrapKey(
+    encryptionKeyId: string,
+    algorithm: string,
+    wrappedKey: Buffer,
+  ): Promise<Buffer> {
+    const origin = this.getOrigin(encryptionKeyId);
+    const keyClient = new KeyClient(origin, this.credentials);
+    const [keyName, keyVersion] = this.getKeyDetails(encryptionKeyId);
+    const cryptographyClient = keyClient.getCryptographyClient(keyName, {
+      keyVersion: keyVersion ? keyVersion : "",
+    });
+    const res = await cryptographyClient.unwrapKey(algorithm as KeyWrapAlgorithm, wrappedKey);
+    return Buffer.from(res.result);
```

Comment:
same as above.

Reply:
I have put a try catch around it, if there are any errors, we don't have any way to recover from there. It is reflected to the client directly.